Add the article for Aurora's dev portal (#1)
mooori/counter-acl-example: Article.md

## Case study: A counter managing permissions with ACL

Let’s look at a case study to see how `near-plugins` can be useful to smart contract developers. We are building a `Counter` which stores its current value and has methods to increment, decrement and reset the value. It is intentionally kept simple to allow us to focus on how `near-plugins` adds functionality. This is what the contract looks like prior to using any plugins:

```rust
#[near_bindgen]
impl Counter {
    #[init]
    pub fn new() -> Self {
        Self { value: 0 }
    }

    /// Anyone can retrieve the current value.
    pub fn value(&self) -> i64 {
        self.value
    }

    /// Increases the value of the counter by one.
    pub fn increment(&mut self) {
        self.value += 1;
    }

    /// Decreases the value of the counter by one.
    pub fn decrement(&mut self) {
        self.value -= 1;
    }

    /// Resets the value of the counter to zero.
    pub fn reset(&mut self) {
        self.value = 0;
    }
}
```

> *The final version of the code is available [in this repository on github](https://github.com/mooori/counter-acl-example). The `Counter` example is inspired by [near-examples/counter-rust](https://github.com/near-examples/counter-rust).*

Reviewer: nit: Will this repo live in your personal github account or do you plan on eventually transferring it to the near or aurora or some other org?  I am asking as we may need to update the link here.  I wonder if it is possible to use relative paths here instead so this question is mute.

Author: Made the links relative in [1582d56](https://github.com/mooori/counter-acl-example/pull/1/commits/1582d565e8d8d623282a9b7ad5c2506055a0ff68).

I’ve created the repo in my personal github as I wasn’t sure were it should live, to avoid triggering some workflows in a potentially wrong org. Moving it to an org is fine for me. Especially if this should be an official example/reference moving it to an org would make sense.

Reviewer: Thanks.  I think we can keep it here for now and figure out eventually which org to place it in.  Longer term, it should probably be in some org.

### Permissions

The contract methods defined above can be called by anyone since they are public and inside an implementation block marked with `#[near_bindgen]`. Using `near-sdk-rs` it is possible to restrict some methods such that they can be called only by the contract itself. Either by using `#[private]` or by not exposing the method publicly, as described in the [documentation](https://docs.near.org/sdk/rust/contract-interface/private-methods).

> However, what if some accounts other than the contract should be able to call restricted methods? This is were [`AccessControllable`](https://github.com/aurora-is-near/near-plugins#accesscontrollable) plugin comes in handy.

Reviewer: ```suggestion
However, what if we wanted to implement more flexible permissions (e.g. allowing some set of accounts to call a given function)? This is where the [`AccessControllable`](https://github.com/aurora-is-near/near-plugins#accesscontrollable) plugin comes in handy.
```

### Managing permissions with ACL

ACL stands for *access control lists* and they are used in the following way within the `AccessControllable` plugin. The user defines the roles required for their use case as Rust enum variants. Then it is possible to restrict access to a method to accounts that have been granted roles. Restricting access is possible with one line of code, for example:

```rust
#[access_control_any(roles(Role::Decrementer))]
pub fn decrement(&mut self) {
    // ...
}
```
 
Let’s walk through it step by step to see how you can make your NEAR smart contract `AccessControllable`.

> ### Stop 1: Add `near-plugins` as a dependency

Reviewer: ```suggestion
### Step 1: Add `near-plugins` as a dependency
```

For now, `near-plugins` has not yet been published on crates.io. Still, the crate is ready for usage and it can be added as git dependency:

```toml
# Add `near-plugins` under `dependencies` in your Cargo.toml.

[dependencies]

> near-plugins = { git = "https://github.com/aurora-is-near/near-plugins.git", rev = "102d95b" }

Reviewer: Do we need to specify a `rev`?  That might mean that people could miss bugfixes, etc. easily.  I suppose this would be resolved when we have released the crate.  Maybe in the mean time, we can use tags?  I am afraid of too many people hardcoding a specific revision in their repo and then forgetting to update.

Author: I agree that this would be resolved when the crate is released and becomes a “regular” dependency. Unless we manually move the tag  as `near-plugins` is changed, I think it would basically be the same: a reference to a commit.

Still, using a tag to the latest release helps explaining why that particular commit is used. It’s about time to release a new version of `near-plugins`. I’ll look into that and then replace the `rev` here with a tag to the new version.

Reviewer: Perfect, thanks, that makes sense.

Author: The [new release](https://github.com/aurora-is-near/near-plugins/releases/tag/v0.2.0) of `near-plugins` has been published. This repo now specifies the `near-plugins` dependency with `tag = "v0.2.0"` (cc1c75e).

```rust
/// See `AccessControllable::acl_grant_role` for details.
acl_grant_role("Resetter", "alice.near");
```

The `AccessControllable` trait provides many more methods to administer ACL permissions. After following the steps above, all of them are automatically implemented for a contract using the `AccessControllable` plugin.

# Done!

The steps above are sufficient to add complex and configurable ACL permissions to a contract using `near-plugins`. At this point, `alice.near` is the only account which has been granted the `Resetter` role. This means that only `alice.near` may successfully call the contract’s `reset()` method.

The repo contains an [integration test](https://example.com) which verifies that `AccessControllable` was setup up correctly for our `Counter` contract. Take a look at it to learn more about interacting with an `AccessControllable` contract. To run the test on-chain in a local sandbox, it suffices to clone the repo and execute the following command. This is made possible by [near-workspaces-rs](https://github.com/near/near-workspaces-rs).

```
cargo test

> ```

Author: This integration test is currently still WIP. I'll give an update once it's ready.

Author: The integration test is now available [here](https://github.com/mooori/counter-acl-example/blob/main/tests/workflow.rs).

## Teaser: How it works internally

Using `AccessControllable` extends the contracts state to store the permissions that have been granted. Moreover, the `AccessControllable` trait is implemented for the contract to enable administering permissions. When `#[access_control_any(roles(...))]` is attached to a method, `near-plugin` injects code which checks if the caller was granted any of the required roles. If not, a panic is generated which aborts the function call.

To learn about all the details, you can dive into the [implementation](https://github.com/aurora-is-near/near-plugins/blob/master/near-plugins-derive/src/access_controllable.rs) of the `AccessControllable` macro.

> ## A note on testing

Reviewer: ```suggestion

## A note on testing
```

The functionality provided by `near-plugins` is critical for security and we strive to test it exhaustively. In tests we compile demo contracts for all plugins and deploy them on-chain in a local sandbox. Then we verify that using a particular plugin adds exactly the expected functionality to the contract. These tests and demo contracts can be found [here](https://github.com/aurora-is-near/near-plugins/tree/master/near-plugins-derive/tests).
## Ready for production, though?

> As mentioned earlier, `near-plugins` comes with the caveat of not yet being published to crates.io. Nevertheless, we are confident that `near-plugins` can already be used in production. In fact, smart contracts related to the [Rainbow Bridge](https://rainbowbridge.app/transfer) already use `near-plugins`. Moreover, both [Hacken](https://www.datocms-assets.com/50156/1680101850-hacken-near-plugins-final-report-updated-march2023.pdf) and [AuditOne](https://www.datocms-assets.com/50156/1680590522-auditone-near-plugins-final-report-updated-march2023.pdf) audited `near-plugins`, awarding it high scores.

Reviewer: > we are confident that `near-plugins` can already be used in production

Hmm... I am a bit vary of the legal implications of such a statement.  I am not a lawyer, but still ca we make this less strong maybe?

Reviewer: How about something like: As mentioned earlier, `near-plugins` comes with the caveat of not yet being published to crates.io.  It is currently used in some contracts on the mainnet e.g. ....

Author: Adapted it according to your proposal in 40e153b.
